pennylane/optimize/shot_adaptive.py

```diff
@@ -0,0 +1,449 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Shot adaptive optimizer"""
+# pylint: disable=too-many-instance-attributes,too-many-arguments,too-many-branches
+from scipy.stats import multinomial
+
+import pennylane as qml
+from pennylane import numpy as np
+
+
+from .gradient_descent import GradientDescentOptimizer
+
+
+class ShotAdaptiveOptimizer(GradientDescentOptimizer):
+    r"""Optimizer with adaptive shot rate, via calculation of the variances of the parameter-shift
+    gradient.
+
+    By keeping a running average of the parameter-shift gradient and the *variance*
+    of the parameter-shift gradient, this optimizer frugally distributes a shot
+    budget across the partial derivatives of each parameter.
+
+    In addition, if computing the expectation value of a Hamiltonian using
+    :class:`~.ExpvalCost`, weighted random sampling can be used to further
+    distribute the shot budget across the terms in the Hamiltonian.
+
+    .. note::
+
+        The shot adaptive optimizer only supports single QNodes or :class:`~.ExpvalCost` objects as
+        objective functions.

```

Comment:
I toyed around with this, the reason I didn't in the end:

* I wasn't 100% sure if everything was well defined if you pass an arbitrary hybrid cost function, since the original proof is written in terms of purely quantum gradients
* It gets a bit complicated if the hybrid cost function contains multiple QNodes, each with different devices

My compromise was:

As an undocumented advanced feature, however, the optimizer will accept a cost function that provides a `.device` attribute.

E.g.,

```python
def cost(x):
    return np.sin(qnode(x))

cost.device = dev
params = opt.step(cost, params)  # this will now work
```

Reply:
But happy to be convinced either way! It was something I was very on the fence about.

Reply:
Your reasoning makes sense, as usual :)